### PR TITLE
refactor: rewrite resources-getter to use query options generator

### PR DESCRIPTION
**Pull request by romain-gilliotte**

### CODE REVIEW

File: src/services/query-builder.js [hunk 1]
```diff
@@ -38,7 +38,7 @@ function getMandatoryFields(association) {
 }
 
 function QueryBuilder(model, opts, params) {
-  const schema = Schemas.schemas[model.name];
+  const schema = Schemas.schemas[model?.name];
```
Comment (romain-gilliotte) [inline src/services/query-builder.js:42]: This will be gone on the last PR (cleanup)

File: src/services/query-options.js [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+
+    const options = { Sequelize: this._Sequelize };
+    const fieldNames = this._requestedFields.size ? [...this._requestedFields] : null;
+    const helper = new SearchBuilder(this._model, options, { search, searchExtended }, fieldNames);
+
+    const conditions = helper.performWithSmartFields(this._options.tableAlias);
+    if (conditions) {
+      this._where.push(conditions);
+    } else {
+      this._where.push(this._Sequelize.literal('(0=1)'));
```
Comment (VincentMolinie) [inline src/services/query-options.js:148]: WHy ? Not sure to understand why you are doing that. This is an OR if I remember so this condition is useless and just add confusion when reading the code

File: src/services/query-options.js [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+    }
+
+    return helper.getFieldsSearched();
+  }
+
+  /**
+   * Apply a forestadmin segment
+   * @param {string} name name of the segment (from the querystring)
+   * @param {string} segmentQuery SQL query of the segment (also from querystring)
```
Comment (VincentMolinie) [inline src/services/query-options.js:157]: segmentQuery ? There is no such params 🤔

File: src/services/query-options.js [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+
+      const associations = await filterParser.getAssociations(filters);
+      associations.forEach(this._neededFields.add, this._neededFields);
+    }
+  }
+
+  /**
+   * Retrict rows to those matching a search string
+   * @param {string} search search string
+   * @param {boolean|string} searchExtended if truthy, enable search in relations
```
Comment (VincentMolinie) [inline src/services/query-options.js:135]: Keep to boolean. Don't encourage people to pass a strring while it's expecting a boolean

File: src/services/query-options.js [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+
+    // ... or as a function which returns a sequelize where clause ...
+    if (typeof segment?.where === 'function') {
+      this._where.push(await segment.where());
+    }
+  }
+
+  /**
+   * Apply a segment query.
+   * FIXME Select SQL injection allows to fetch any information from database.
```
Comment (VincentMolinie) [inline src/services/query-options.js:177]:
```suggestion
   * FIXME: Select SQL injection allows to fetch any information from database.
```

File: src/services/query-options.js [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+   * @param {string} sortString a sort string
+   */
+  async sort(sortString) {
+    if (!sortString) {
+      return;
+    }
+
+    let order = 'ASC';
+    if (sortString[0] === '-') {
+      sortString = sortString.substring(1);
```
Comment (VincentMolinie) [inline src/services/query-options.js:213]: Did you bypass the linter :o ? This is not allowed. You can't reassign a parameter

File: src/services/query-options.js [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+      order = 'DESC';
+    }
+
+    if (sortString.indexOf('.') !== -1) {
```
Comment (VincentMolinie) [inline src/services/query-options.js:217]:
```suggestion
    if (sortString.includes('.')) {
```

File: src/services/resources-getter.js [hunk 8]
```diff
@@ -1,255 +1,78 @@
...
-    this.model = model;
-    this.options = options;
-    this.params = params;
-    this.schema = Schemas.schemas[model.name];
-    this.queryBuilder = new QueryBuilder(model, options, params);
-    this.operators = Operators.getInstance(options);
-    [this.primaryKey] = _.keys(model.primaryKeys);
-    this.filterParser = new FiltersParser(this.schema, params.timezone, options);
+  constructor(model, lianaOptions, params) {
+    // The liana options argument is kept for retro-compatibility.
```
Comment (VincentMolinie) [inline src/services/resources-getter.js:9]: Retrocompatibility but you are not deprecating the old way of instantiating. I think you should

File: src/services/resources-getter.js [hunk 9]
```diff
@@ -1,255 +1,78 @@
...
-          // NOTICE: No search condition has been set for the current search, no record can be
-          //         found.
-          return 0;
-        }
-      }
-    }
-
-    return scope.count(queryOptions);
+  /** Load records matching current query (w/ pagination) */
+  async _getRecords() {
```
Comment (VincentMolinie) [inline src/services/resources-getter.js:33]: AVoid reordering method on refacto, it increase the diff and it's way harder for the reviewer to see the difference

**Review by VincentMolinie — CHANGES_REQUESTED** [review]

File: src/services/query-options.js [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+  }
+
+  /**
+   * Add the required includes from a list of field names.
+   * @param {string[]} fields Fields of HasOne and BelongTo relations are
+   *  accepted (ie. 'book.name').
+   */
+  async requireFields(fields) {
+    if (fields) {
+      fields.forEach(this._requestedFields.add, this._requestedFields);
```
[the remark this answers is not in the document]
Comment (romain-gilliotte) [reply src/services/query-options.js:102]: the second argument of forEach is the function you want to bind to
This is equivalent to 
`fields.forEach(this._requestedFields.add.bind(this._requestedFields));`

File: src/services/query-options.js [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+   */
+  async filterByConditionTree(filters, timezone) {
+    const filterParser = new FiltersParser(this._schema, timezone, { Sequelize: this._Sequelize });
+
+    if (filters) {
+      const whereClause = await filterParser.perform(filters);
+      this._where.push(whereClause);
+
+      const associations = await filterParser.getAssociations(filters);
+      associations.forEach(this._neededFields.add, this._neededFields);
```
[the remark this answers is not in the document]
Comment (romain-gilliotte) [reply src/services/query-options.js:128]: same explanation

File: src/services/query-options.js (hunk 2 above)
Comment (romain-gilliotte) [reply src/services/query-options.js:148]: This is a "and".
This line ensures that the query options will not return any results at all.

In the former code it replaces this patch, which was solving a search related bug:
- https://github.com/ForestAdmin/forest-express-sequelize/blob/master/src/services/resources-getter.js#L157-L166
- https://github.com/ForestAdmin/forest-express-sequelize/blob/master/src/services/resources-getter.js#L204-L213

File: src/services/query-options.js (hunk 3 above)
Comment (romain-gilliotte) [reply src/services/query-options.js:157]: Sadly, there is such a param, and it's a SQL injection

https://github.com/ForestAdmin/forest-express-sequelize/blob/master/src/services/resources-getter.js#L95-L98

File: src/services/query-options.js (hunk 4 above)
Comment (romain-gilliotte) [reply src/services/query-options.js:135]: I'm passing the query param directly from the resources-getter.
I'll change the jsdoc

File: src/services/query-options.js [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+    this._where.push(new PrimaryKeysManager(this._model).getRecordsConditions(recordIds));
+  }
+
+  /**
+   * Apply condition tree to those query options (scopes, user filters, charts, ...)
+   * @param {*} filters standard forest filters
+   * @param {*} timezone timezone of the user (required if filtering on dates)
+   */
+  async filterByConditionTree(filters, timezone) {
+    const filterParser = new FiltersParser(this._schema, timezone, { Sequelize: this._Sequelize });
```
[the remark this answers is not in the document]
Comment (romain-gilliotte) [reply src/services/query-options.js:121]: i'll early quit the function like in the others

File: src/services/query-options.js (hunk 5 above)
Comment (romain-gilliotte) [reply src/services/query-options.js:177]: ok

File: src/services/query-options.js (hunk 6 above)
Comment (romain-gilliotte) [reply src/services/query-options.js:213]: Looks like the linter configuration is not the same in frontend and this project: the linter did not complain on this one.

File: src/services/resources-getter.js (hunk 8 above)
Comment (romain-gilliotte) [reply src/services/resources-getter.js:9]: This class is instanciated from another project (forest-express).
I don't want to release a new forest-express version => that's why i kept this parameter which is now useless

File: src/services/query-options.js (hunk 7 above)
Comment (romain-gilliotte) [reply src/services/query-options.js:217]: I'll change that.
I copy pasted it from the old code :)

File: src/services/resources-getter.js (hunk 9 above)
Comment (romain-gilliotte) [reply src/services/resources-getter.js:33]: There is no difference in this file.
I deleted the old one and rewrote it from scratch...

File: src/services/resources-getter.js (hunk 8 above)
Comment (romain-gilliotte) [reply src/services/resources-getter.js:9]: added test to be able to instanciate the new way

File: src/services/query-options.js (hunk 3 above)
Comment (VincentMolinie) [reply src/services/query-options.js:156]: then it shoul be a single param called nameOrSegmentQuery ?

**Review by VincentMolinie — APPROVED** [review]

**Pull request merged**